rpc/client.go

```diff
@@ -91,219 +97,224 @@ type Client struct {
 //
 // The Client.Ready channel is closed after the client has connected
 // and completed one successful heartbeat. The Closed channel is
-// closed if the client fails to connect or if the client's Close()
-// method is invoked.
-func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
+// closed if the client's Close() method is invoked.
+func NewClient(addr net.Addr, context *Context) *Client {
 	clientMu.Lock()
+	defer clientMu.Unlock()
+
+	key := util.MakeUnresolvedAddr(addr.Network(), addr.String())
+
 	if !context.DisableCache {
-		if c, ok := clients[addr.String()]; ok {
-			clientMu.Unlock()
+		if c, ok := clients[key]; ok {
 			return c
 		}
 	}
+
+	tlsConfig, err := context.GetClientTLSConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
 		Closed:       make(chan struct{}),
+		addr:         key,
+		tlsConfig:    tlsConfig,
 		clock:        context.localClock,
 		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
-	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
 	}
-	clientMu.Unlock()
 
-	context.Stopper.RunWorker(func() {
-		if err := c.connect(opts, context); err != nil {
-			log.Errorf("client %s failed to connect: %v", c.addr, err)
-		} else {
-			// Signal client is ready by closing Ready channel.
-			log.Infof("client %s connected", c.addr)
-			close(c.Ready)
-
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
-		c.Close()
-	})
-	return c
-}
+	c.healthy.Store(make(chan struct{}))
+	// Must store junk in here so `Load()` doesn't blow up
+	c.conn.Store(internalConn{})
+
+	clients[key] = c
```

Comment:
We should not cache clients that were created with `context.DisableCache=true` (because in the case of gossip they have an incorrect value of `context.RemoteClocks`), unless we can consider the context as a part of the cache key.